mongodb/src/main/java/org/hibernate/ogm/datastore/mongodb/index/MongoDBIndexSpec.java

```diff
@@ -0,0 +1,166 @@
+/*
+ * Hibernate OGM, Domain model persistence for NoSQL datastores
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.ogm.datastore.mongodb.index;
+
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.Map;
+
+import org.hibernate.mapping.Column;
+import org.hibernate.mapping.Index;
+import org.hibernate.mapping.UniqueKey;
+import org.hibernate.ogm.datastore.mongodb.options.impl.MongoDBIndexOptions;
+import org.hibernate.ogm.datastore.mongodb.options.impl.MongoDBTextIndexOptions;
+import org.hibernate.ogm.index.OgmIndexSpec;
+import org.hibernate.ogm.util.impl.StringHelper;
+
+import com.mongodb.BasicDBObject;
+import com.mongodb.DBObject;
+
+/**
+ * Definition of an index to be applied to a MongoDB collection
+ *
+ * @author Francois Le Droff
+ * @author Guillaume Smet
+ */
+public class MongoDBIndexSpec implements OgmIndexSpec {
+
+	/**
+	 * The MongoDB collection/table for which the index will be set
+	 */
+	private String collection;
+
+	/**
+	 * A unique index causes MongoDB to reject all documents that contain a duplicate value for the indexed field.
+	 * @see http://docs.mongodb.org/manual/core/index-unique/
+	 * true if the indexed field is bound to be unique within the collection
+	 */
+	private boolean unique;
+
+	/**
+	 * Optional. The name of the index. If unspecified, MongoDB generates an index name
+	 * by concatenating the names of the indexed fields and the sort order.
+	 *
+	 * Whether user specified or MongoDB generated, index names including their full namespace (i.e. database.collection)
+	 * cannot be longer than the Index Name Limit (that is 128 characters)
+	 */
+	private String indexName;
+
+	/**
+	 * The index keys of the index prepared for MongoDB.
+	 */
+	private DBObject indexKeys = new BasicDBObject();
+
+	/**
+	 * The options specific to MongoDB.
+	 */
+	private MongoDBIndexOptions options;
+
+	public MongoDBIndexSpec(UniqueKey uniqueKey, MongoDBIndexOptions options) {
+		this.options = options;
+		this.collection = uniqueKey.getTable().getName();
+		this.indexName = uniqueKey.getName();
+		this.unique = true;
+		this.addIndexKeys( uniqueKey.getColumnIterator(), uniqueKey.getColumnOrderMap() );
+	}
+
+	public MongoDBIndexSpec(Index index, MongoDBIndexOptions options) {
+		this.options = options;
+		this.collection = index.getTable().getName();
+		this.indexName = index.getName();
+		this.unique = false;
+		// TODO HHH-10760: the columnOrderMap is not accessible for an Index
```

Comment:
Done: https://hibernate.atlassian.net/browse/OGM-1080 . I referenced it in the TODO.